File: clients/android/sample/src/androidTest/kotlin/dev/arizona/sample/MenuE2ETest.kt
package dev.arizona.sample

import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import org.junit.Rule
import org.junit.Test

/**
 * On-device navigation: from the menu, tap a button to navigate to another view
 * (same-socket transition -> OP_REPLACE), then the chrome "Menu" button to
 * navigate back. Run with ./gradlew :sample:connectedCheck (server at
 * 10.0.2.2:4040).
 */
class MenuE2ETest {
    @get:Rule
    val rule = createAndroidComposeRule<MainActivity>()

    @Test
    fun navigatesToViewAndBack() {
        rule.waitForText("Counter")
        // Forward: menu -> counter.
        rule.onNodeWithText("Counter").performClick()
        rule.waitForText("Count: 0")

        // Back: chrome "Menu" button -> the menu re-mounts.
        rule.onNodeWithText("☰ Menu").performClick()
        rule.waitForText("Counter")
    }
}

File: clients/ios/Sample/UITests/MenuE2ETest.swift
import XCTest

/// On-device navigation: from the menu, navigate to a view (same-socket
/// transition -> OP_REPLACE), then the chrome "Menu" button back. The Android
/// analogue is MenuE2ETest.
final class MenuE2ETest: XCTestCase {
    private var app: XCUIApplication!

    override func setUpWithError() throws {
        continueAfterFailure = false
        app = launchSample()
    }

    func testNavigatesToViewAndBack() {
        // Forward: menu -> counter.
        tapButton(app, "Counter")
        waitForText(app, "Count: 0")

        // Back: chrome "Menu" button -> the menu re-mounts.
        tapButton(app, "\u{2630} Menu")
        waitForText(app, "Counter")
    }
}
